### PR TITLE
add minimum key size checks and warnings

### PR DESCRIPTION
**Pull request by mattsb42-aws**
*Issue #, if available:* #78 

*Description of changes:*
We now log a warning if you create or load an HMAC key shorter than 128 bits or a RSA key shorter than 2048 bits.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mattsb42-aws** [comment]
Note: This is still unmerged because something about the tests I added are making Travis fail... I'll be looking into that this week.

**Comment by mattsb42-aws** [comment]
rebased on top of recent changes; still trying to figure out why this kills Travis...

**Comment by mattsb42-aws** [comment]
Changes made to:
1. Disable these tests on Travis because for some reason they are causing Travis to crash. As we discussed, I am going to focus on getting us off of Travis rather than spending more time diving down this rabbit hole trying to figure out why Travis is crashing.
2. Re-enable examples tests that became broken at some point. We don't currently run these in CI because they need to interact with a real DDB table.

**Comment by mattsb42-aws** [comment]
...looks like Travis is still crashing, which tells me it's probably the helpers I added, not the actual tests...for the short-term fix I think I'll try just dropping the Travis runs to the "fast" runs instead of "slow"...

File: test/functional/delegated_keys/test_jce.py [hunk 1]
```diff
@@ -49,3 +55,28 @@ def test_generate_correct_key_length(algorithm, requested_bits, expected_bits, l
...
+    yield ("RSA", MinimumKeySizes.RSA.value, False, message)
+    yield ("RSA", MinimumKeySizes.RSA.value // 2, True, message)
+
+
+@pytest.mark.parametrize("algorithm, key_bits, too_short, error_message", build_short_key_cases())
+def test_warn_on_short_keys(capturing_logger, algorithm, key_bits, too_short, error_message):
+    _test = JceNameLocalDelegatedKey.generate(algorithm, key_bits)
+
+    logging_results = capturing_logger.getvalue()
+    assert (too_short and error_message in logging_results) or (not too_short and error_message not in logging_results)
```
Comment (karlw00t) [inline test/functional/delegated_keys/test_jce.py]: I'm not going to block on this, but nested list comprehensions can make this code hard to parse.

File: test/functional/functional_test_utils.py [hunk 2]
```diff
@@ -641,3 +645,13 @@ def client_cycle_batch_items_check_paginators(
     e_scan_result = e_client.scan(TableName=table_name, ConsistentRead=True)
     assert not raw_scan_result["Items"]
     assert not e_scan_result["Items"]
+
+
+@pytest.fixture
+def capturing_logger():
```
Comment (karlw00t) [inline test/functional/functional_test_utils.py]: Why can't we use Capture Log here?

https://pypi.org/project/pytest-capturelog/

File: test/functional/delegated_keys/test_jce.py (hunk 1 above)
Comment (mattsb42-aws) [reply test/functional/delegated_keys/test_jce.py]: @karlw00t That's not a list comprehension, it's just two boolean checks

It just compressed this:
```python
if too_short:
    assert error_message in logging_results
else:
    assert error_message not in logging_results
```
to this:
```python
assert (too_short and error_message in logging_results) or (not too_short and error_message not in logging_results)
```

File: test/functional/functional_test_utils.py (hunk 2 above)
Comment (mattsb42-aws) [reply test/functional/functional_test_utils.py]: Looks like they don't document it, but yes, we can use that.

The undocumented bit is that you can actually get the logged data[1].

[1] https://bitbucket.org/memedough/pytest-capturelog/src/039cb3be01c9c6480d3907c47247f5bb0744eb36/pytest_capturelog.py?at=default&fileviewer=file-view-default#pytest_capturelog.py-233:236

Comment (mattsb42-aws) [reply test/functional/functional_test_utils.py]: I thought I remembered this being somewhere else...turns out this is part of pytest now.

https://docs.pytest.org/en/latest/logging.html

**Comment by mattsb42-aws** [comment]
In updating to use pytest's built-in `caplog` fixture, I updated the pytest dependency minimum version because that was the last version that the `caplog` module was changed. This necessitated an update in the upstream test requirements.
